src/final_chain/final_chain.cpp

```diff
@@ -0,0 +1,440 @@
+    return state_api.dpos_query(last_if_absent(blk_n), q);
+  }
+
+  EthBlockNumber last_if_absent(optional<EthBlockNumber> const& client_blk_n) const {
+    return client_blk_n ? *client_blk_n : last_block_number();
+  }
+
+  static util::RangeView<state_api::EVMTransaction> to_state_api_transactions(Transactions const& trxs) {
+    return make_range_view(trxs).map([](auto const& trx) {
+      return state_api::EVMTransaction{
+          trx.getSender(), trx.getGasPrice(), trx.getReceiver(), trx.getNonce(),
+          trx.getValue(),  trx.getGas(),      trx.getData(),
+      };
+    });
+  }
+
+  BlocksBlooms block_blooms(h256 const& chunk_id) const {
+    if (auto raw = db_->lookup(chunk_id, DB::Columns::final_chain_log_blooms_index); !raw.empty()) {
+      return RLP(raw).toArray<LogBloom, c_bloomIndexSize>();
+    }
+    return {};
+  }
+
+  static h256 block_blooms_chunk_id(EthBlockNumber level, EthBlockNumber index) { return h256(index * 0xff + level); }
+
+  vector<EthBlockNumber> withBlockBloom(LogBloom const& b, EthBlockNumber from, EthBlockNumber to, EthBlockNumber level,
+                                        EthBlockNumber index) const {
+    vector<EthBlockNumber> ret;
+    auto uCourse = int_pow(c_bloomIndexSize, level + 1);
+    auto uFine = int_pow(c_bloomIndexSize, level);
+    auto obegin = index == from / uCourse ? from / uFine % c_bloomIndexSize : 0;
+    auto oend = index == to / uCourse ? (to / uFine) % c_bloomIndexSize + 1 : c_bloomIndexSize;
+    auto bb = block_blooms(block_blooms_chunk_id(level, index));
+    for (auto o = obegin; o < oend; ++o) {
+      if (bb[o].contains(b)) {
+        // This level has something like what we want.
+        if (level > 0) {
+          dev::operator+=(ret, withBlockBloom(b, from, to, level - 1, o + index * c_bloomIndexSize));
+        } else {
+          ret.push_back(o + index * c_bloomIndexSize);
+        }
+      }
+    }
+    return ret;
+  }
+
+  struct TransactionHashesImpl : TransactionHashes {
+    string serialized_;
+    size_t count_;
+
+    explicit TransactionHashesImpl(string serialized)
+        : serialized_(move(serialized)), count_(serialized_.size() / h256::size) {}
+
+    static bytes serialize_from_transactions(Transactions const& transactions) {
+      bytes serialized;
+      serialized.reserve(transactions.size() * h256::size);
+      for (auto const& trx : transactions) {
+        for (auto b : trx.getHash()) {
+          serialized.push_back(b);
+        }
+      }
+      return serialized;
+    }
+
+    h256 get(size_t i) const override {
+      return h256((uint8_t*)(serialized_.data() + i * h256::size), h256::ConstructFromPointer);
+    }
+
+    size_t count() const override { return count_; }
+  };
+};
+
+unique_ptr<FinalChain> NewFinalChain(shared_ptr<DB> const& db, Config const& config, Opts const& opts,
+                                     addr_t const& node_addr) {
+  return u_ptr(new FinalChainImpl(db, config, opts, node_addr));
```

Comment:
do not use "new", use make_shared/unique instead

Reply:
ok about make_unique.

>What is this whole NewFinalChain function even good for ?

It is the only publicly exposed thing in this .cpp file. The idea of this pattern is to completely hide the implementation. The header defines an interface + a factory (this function). And the only job .cpp file does - provide an interface implementation by defining the factory.

Reply:
ok for factory function

Reply:
done with the `new`